Cache unitless force parameters to increase performance

bas-rustenburg opened the pull request:
This PR is an attempt to speed up the update function by removing units. Work in progress.

jchodera commented:
I love the unit stripper, but I am a bit worried about this approach to stripping units since it could be dangerous if we mix stripped and non-stripped units or somehow get the units in different unit systems.

If the time-consuming code is in `setTitrationState`, why don't we simply modify this function to do the following:
* On the first call, it memoizes (caches) stripped versions of the charges it needs to remember
* It always uses the cached stripped version of the charges to update charges and exceptions

That is all we need to do to eliminate the performance bottleneck arising from unit conversions.  We can keep units everywhere else without risking the same kind of insanity affecting perses.

swails commented on constph.py:
```diff
@@ -55,9 +55,17 @@
+    """Strips the unit from a simtk.units.Quantity object and returns it's value in OpenMM consistent units.
+        Returns quant if object is not a quantity."""
+    if units.is_quantity(quant):
+        return quant.in_unit_system(units.md_unit_system)._value
```
You can replace this with `value_in_unit(units.md_unit_system)`.  Probably better than using the `_value` attribute.  The only time I use `_value` directly is when I want to keep the number as a unit, but explicitly bypass dimensional analysis.

swails commented on constph.py:
```diff
@@ -55,9 +55,17 @@
 import simtk.unit as units
 import pymbar
 
+def strip_unit(quant):
```
For added safety, I would recommend changing the call signature to

```Python
def strip_unit(quant, unit):
```

Where `unit` is the target unit.  That way, you can properly throw an exception when the input units are incompatible with what's expected.  This is a potential cause of bugs.  For instance, one of my friends once instantiated a LangevinIntegrator as such:

```
integrator = LangevinIntegrator(1/femtosecond, 300*kelvin)
```

only to have the simulation immediately segfault because the time step was expanded to 2000/picosecond instead of 0.002 picosecond (and the unit implicitly dropped).  The only reason we can't use this approach in OpenMM is because the code is embedded in the SWIG/C++ translation layer, and it's a real challenge to accommodate everywhere that it's needed.

jchodera commented on constph.py:
```diff
@@ -55,9 +55,17 @@
 # MODULE CONSTANTS
 kB = units.BOLTZMANN_CONSTANT_kB * units.AVOGADRO_CONSTANT_NA
-kB = kB.in_units_of(units.kilocalories_per_mole/units.kelvin)
+kB = strip_unit(kB)
```
We don't want to strip units globally.  Please don't do this.  

We just want to avoid the unit-stripping overhead in the regions of the code that are called over and over again, which are only three lines in `setTitrationState`.  We can do this by caching the stripped values.

bas-rustenburg commented:
Reverted some changes. Will try and implement the caching so we only use those unitless charges once and never change them.

bas-rustenburg commented:
@swails, made the unit stripping more foolproof based on some of your suggestions.

bas-rustenburg commented:
Implemented an option to precalculate all force parameters in the __init__ function.

Increases the setup cost, but a quick look at a single amino acid seems to lead to a 3-10x speedup for titration trials, though I'll run a proper benchmark as well.

bas-rustenburg commented:
Diff gets rendered a little weird, best to look at the full file.

jchodera commented on constph.py:
```diff
@@ -573,14 +618,14 @@ def setTitrationState(self, titration_group_index, titration_state_index, contex
-                    [charge, sigma, epsilon] = force.getParticleParameters(atom_index)
-                    if debug: print (" modifying NonbondedForce atom %d : (charge, sigma, epsilon) : (%s, %s, %s) -> (%s, %s, %s)" % (
-                    atom_index, str(charge), str(sigma), str(epsilon), str(charges[charge_index]), str(sigma), str(epsilon)))
+                    [charge, sigma, epsilon] = map(strip_in_unit_system, force.getParticleParameters(atom_index))
```
Don't we want to cache `[charge, sigma, epsilon]` to avoid the unit conversion overhead?  Ideally, we want to avoid calling `getParticleParameters`, which is also slow.

jchodera commented on constph.py:
```diff
@@ -573,14 +618,14 @@ def setTitrationState(self, titration_group_index, titration_state_index, contex
-                    [charge, radius, scaleFactor] = force.getParticleParameters(atom_index)
+                    # if debug: print (" modifying GBSAOBCForce atom %d : (charge, radius, scaleFactor) : (%s, %s, %s) -> (%s, %s, %s)" % (
+                    # atom_index, str(charge), str(radius), scaleFactor, str(charges[charge_index]), str(radius), scaleFactor))
+                    [charge, radius, scaleFactor] = map(strip_in_unit_system, force.getParticleParameters(atom_index))
```
We want to avoid calling `getParticleParameters` by caching `[charge, radius, scaleFactor]`.

jchodera commented on constph.py:
```diff
@@ -588,28 +633,116 @@ def setTitrationState(self, titration_group_index, titration_state_index, contex
-                    [charge1, sigma1, epsilon1] = force.getParticleParameters(particle1)
-                    [charge2, sigma2, epsilon2] = force.getParticleParameters(particle2)
-                    #print "chargeProd: old %s new %s" % (str(chargeProd), str(self.coulomb14scale * charge1 * charge2))
+                    [particle1, particle2, chargeProd, sigma, epsilon] = map(strip_in_unit_system,
```
We want to avoid all of these `get*` calls, since this incurs overhead, by caching all of the parameters we need for `force.setExceptionParameters(...)`.

jchodera commented:
OK, I'm confused how this works.  Maybe you can walk me through it tomorrow?

jchodera commented:
Is the idea that you have two code paths: one optimized and one unoptimized, so you can compare performance?  If so, shouldn't one exclusively use stripoed units, while the other shouldn't strip units at all?

bas-rustenburg commented:
Yes, sorry for the confusion. Might be clearest to look at [`setTitrationState`](https://github.com/choderalab/openmm-constph/blob/stripunits/constph.py#L595-L603). Basically I left the old way intact to compare, but that one currently strips units, without caching them. I was curious about the effect of just stripping units manually, but not getting rid of the get calls. I intend to remove it entirely later, unless we can think of a reason to not precache all of these.

jchodera commented:
What do `_force_updates()` and `_read_cached_force()` do? 

Would be helpful to write docstrings even for non-exported (private) functions!  Unless you never intend anyone to read/use/maintain your code, it is a good idea to document what each thing is supposed to do, even if it's not part of the public API.

jchodera commented:
If `_force_updates()` is supposed to be the old, unoptimized way, we shouldn't be using `strip_in_unit_system`, right?
https://github.com/choderalab/openmm-constph/blob/stripunits/constph.py#L621-L624

And `_read_cached_force()` actually sets the force parameters, rather than reading them?

So confused.

jchodera commented:
I see: `_read_cached_forces` is supposed to read from the *cache* and then set the updated parameters in forces.

jchodera commented:
OK, the changes for caching are spread in a few places.

This is just a bit of a design choice, but it might be good to localize the whole caching framework in `setTitrationState()`.  You can have `__init__` store `self.cache_forces = True`, and then `setTitrationState` can build the cache if it doesn't exist yet before using it.  That keeps all the changes localized.

bas-rustenburg commented:
>Would be helpful to write docstrings even for non-exported (private) functions!

Of course, I decided not to prematurely write them while I was still juggling code. Sorry if that made the development code confusing. I'll add them in the next iteration.

>If _force_updates() is supposed to be the old, unoptimized way, 

I mostly kept the old code around for comparison,  in case I screwed anything up while coding up the cached solution. Though I was curious about the effect of just stripping the units, it doesn't seem to have as much an effect as caching everything, so I don't need to benchmark it. 

> I see: _read_cached_forces is supposed to read from the cache and then set the updated parameters in forces.

Yeah, sorry for the confusion. Names aren't final yet. How about just calling it `_update_forces`. I will remove the old `_force_updates` that was temporarily in place. The fact that forces are cached are just a detail now.

> This is just a bit of a design choice, but it might be good to localize the whole caching framework in setTitrationState(). You can have __init__ store self.cache_forces = True, and then setTitrationState can build the cache if it doesn't exist yet before using it. 

As far as design choices go, I don't like the idea of any preparation steps slipping into the `update` through `setTitrationState`. I feel like `__init__` is the proper place to do this, since it only needs to be called once, and is useful even before the first iteration. I intend to remove the option to cache forces and always cache it in `__init__`. No reason to not speed up the code. Additional benefit is that it'll keep the responsibilities of `setTitrationState` simpler and more isolated.

bas-rustenburg commented:
Cleaned up the code a bit and added docstrings. Let me explain the structure a bit:

Forces are now cached here, as states are added:
https://github.com/choderalab/openmm-constph/blob/stripunits/constph.py#L209
Instead of one function that caches all of them, now it is part of the state definition. 

Caching is implemented here :
https://github.com/choderalab/openmm-constph/blob/stripunits/constph.py#L629

`setTitrationState` simply calls everything that needs to be changed to switch to the new titration state.
https://github.com/choderalab/openmm-constph/blob/stripunits/constph.py#L584

To do so, it calls `_update_forces` , to updates the force parameters from a new titration state, by reading the cache.
https://github.com/choderalab/openmm-constph/blob/stripunits/constph.py#L589

bas-rustenburg commented:
Benchmark results:

| System 	| old step/trial ratio 	| new step/trial ratio 	|
|--------	|----------------------	|----------------------	|
| dhfr   	| 77.209022            	| 26.967738            	|
| abl    	| 39.179890            	| 15.058011            	|
| ldha   	| 4.857071             	| Still running        	|
| his    	| Not measured         	| 126.476722           	|


call graphs:
[dhfr](https://github.com/choderalab/openmm-constph/blob/stripunits/benchmarks/dhfr-implicit/benchmark_cache.png)
[abl](https://github.com/choderalab/openmm-constph/blob/stripunits/benchmarks/abl-implicit/benchmark_cache.png)
[his](https://github.com/choderalab/openmm-constph/blob/stripunits/benchmarks/histidine-implicit/benchmark_cache.png)

See the benchmark_cache.prof files in each folder for the full profile.

jchodera commented:
Great!  If we can merge this, I can start on NCMC switching.

bas-rustenburg commented:
Sure, should be good to merge.
